crates/bevy_reflect/src/serde/ser.rs

```diff
@@ -214,104 +255,114 @@ impl<'a> Serialize for EnumSerializer<'a> {
     where
         S: serde::Serializer,
     {
-        let mut state = serializer.serialize_map(Some(2))?;
-
-        state.serialize_entry(type_fields::TYPE, self.enum_value.type_name())?;
-        state.serialize_entry(
-            type_fields::ENUM,
-            &EnumValueSerializer {
-                enum_value: self.enum_value,
-                registry: self.registry,
-            },
+        let type_info = get_type_info(
+            self.enum_value.get_type_info(),
+            self.enum_value.type_name(),
+            self.registry,
         )?;
-        state.end()
-    }
-}
-
-pub struct EnumValueSerializer<'a> {
-    pub enum_value: &'a dyn Enum,
-    pub registry: &'a TypeRegistry,
-}
 
-impl<'a> Serialize for EnumValueSerializer<'a> {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: serde::Serializer,
-    {
-        let variant_type = self.enum_value.variant_type();
-        let variant_name = self.enum_value.variant_name();
-
-        let mut state = if matches!(variant_type, VariantType::Unit) {
-            serializer.serialize_map(Some(1))?
-        } else {
-            serializer.serialize_map(Some(2))?
+        let enum_info = match type_info {
+            TypeInfo::Enum(enum_info) => enum_info,
+            info => {
+                return Err(Error::custom(format_args!(
+                    "expected enum type but received {:?}",
+                    info
+                )));
+            }
         };
 
-        state.serialize_entry(type_fields::VARIANT, variant_name)?;
+        let enum_name = enum_info.name();
+        let variant_info = enum_info
```

Comment:
Yeah `Enum::variant_index` could work to reduce the hash lookups. I was going to suggest `Enum::variant_info`, but that would also be inefficient as we already have access to info here.

Reply:
Yeah I think `Enum::variant_index()` combined with `EnumInfo::variant_at` is sufficient.

Reply:
I updated the code. One note is that adding `Enum::variant_index` meant that we had to update `DynamicEnum` to account for this. I added `DynamicEnum::new_with_index` and `DynamicEnum::set_variant_with_index`.

However, this change means that we now rely on the index being properly set. If a user tries to serialize `DynamicEnum`, they need to make sure they actually set the index.

We could just force the user to define this whenever they set the variant or create an instance of `DynamicEnum`, but it makes the API a bit clunkier (and you need to know/remember the variant order).

---

**Side note:** For others' reference, I don't think we need to be concerned with the stability of variant order here. This change only affects serialization. So, except in very specific circumstances, the index should always be valid (at least for derived enums).

Reply:
Hm, I'm wondering if I should just revert back to using the previous solution. I'm a little concerned about the potential pitfalls this introduces for users trying to serialize incomplete `DynamicEnum`s (i.e. ones missing a valid variant index).

Reply:
Oh, this also made me realize that this we really rely on proper ordering for things that shouldn't really care (at least in a Dynamic context), such as with structs. Not sure if we want to address that and opt for the name lookups instead (would involve editing `Struct::iter_fields` probably).

That could also (and probably should) just be addressed in a future PR.

Reply:
> Hm, I'm wondering if I should just revert back to using the previous solution. I'm a little concerned about the potential pitfalls this introduces for users trying to serialize incomplete DynamicEnums (i.e. ones missing a valid variant index).

Yeah this problem is the same for DynamicStructs and serialization (which also rely on the dynamic struct's field_at aligning with the actual TypeInfo's field_at). 

Seems like we might want a "fast path" for "real structs/enums" and a "always correct path"  (using field names) for dynamic structs/enums, rather than treating both as the same.

I don't want to revert the fast path (the common case) in the interest of correctness for the (very) uncommon case of using Dynamic types in user code (and then trying to serialize them).

Short term, I'm cool with the current impl as it aligns with the struct pattern, but post-merge we can follow up with an issue / discuss fixes for these cases.